modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpHeaderValidator.java

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.http.netty4;
+
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.handler.codec.DecoderResult;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpHeaderValues;
+import io.netty.handler.codec.http.HttpMessage;
+import io.netty.handler.codec.http.HttpObject;
+import io.netty.handler.codec.http.LastHttpContent;
+import io.netty.util.ReferenceCountUtil;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.action.ActionListener;
+
+import java.util.ArrayDeque;
+import java.util.function.BiConsumer;
+
+public class Netty4HttpHeaderValidator extends ChannelInboundHandlerAdapter {
+
+    private final BiConsumer<HttpMessage, ActionListener<Void>> validator;
+    private ArrayDeque<HttpObject> pending = new ArrayDeque<>(4);
+    private STATE state = STATE.WAITING_TO_START;
+
+    public Netty4HttpHeaderValidator(BiConsumer<HttpMessage, ActionListener<Void>> validator) {
+        this.validator = validator;
+    }
+
+    STATE getState() {
+        return state;
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        assert msg instanceof HttpObject;
+        final HttpObject httpObject = (HttpObject) msg;
+
+        if (state == STATE.WAITING_TO_START) {
+            assert pending.isEmpty();
+            pending.add(ReferenceCountUtil.retain(httpObject));
+            requestStart(ctx);
+        } else if (state == STATE.QUEUEING_DATA) {
+            pending.add(ReferenceCountUtil.retain(httpObject));
+        } else if (state == STATE.HANDLING_QUEUED_DATA) {
+            pending.add(ReferenceCountUtil.retain(httpObject));
+            // Immediately return as this can only happen from a reentrant read(). We do not want to change
+            // autoread in this case.
+            return;
+        } else if (state == STATE.FORWARDING_DATA) {
+            assert pending.isEmpty();
+            if (httpObject instanceof LastHttpContent) {
+                state = STATE.WAITING_TO_START;
+            }
+            ctx.fireChannelRead(httpObject);
```

Comment:
#### Background

Security is used to convey the authentication information (when successful) in thread-local variables that we call the `threadContext`.
In essence, all the code that's run by a thread with some authentication information in the `threadContext` is said to run "as" the "subject" from that authentication information.
Today, the `SecurityRestFilter` sets and clears the `threadContext`, in a try-block, before and after executing the request handler.

#### Question

Assuming authentication is successful, and here we're forwarding `httpObject`s to the next pipeline handler. We need to ensure that the `threadContext` contains the authentication information, when the **complete request** gets to the `RestController`.

Is is correct to set the `threadContext` to contain the authentication when we forward the `LastHttpContent` message?

I think the downstream pipeline handler, upon seeing this message will assemble the complete request and forward it further, **under the same thread call stack**, i.e. not enqueue it, like it does with the other `httpObject`s and wait for some other event to forward the complete request, in which case the `threadContext` will not be populated correctly.

Reply:
Yeah. We cannot propagate this through thread locals at this point. As additional context switching could happen. If there is information from Authentication that we need to propagate, the Validator will have to return it and we can find some mechanism to shove it in the http request object until "handling" where we can set the thread context.

Reply:
> we can find some mechanism to shove it in the http request object until "handling" where we can set the thread context.

We discussed to wrap and decorate the `HttpRequest` that we forward with the authn result. That looks good, but the problem is that the `HttpRequest` is not available from the `FullHttpRequest` that the `RestController` ultimately deals with, so it cannot get hold of the authn result either (in order to set it in the thread context that runs the request handler).
In order to expose the authn result to the `FullHttpRequest` we'll need to extend the `HttpObjectAggregator` slightly (more precisely extend `#beginAggregation` to return an implementation of `FullHttpMessage` that contains the authn result from the `HttpMessage start` parameter). 

I'm going to try implementing this.